### PR TITLE
[OSDEV-1409] Implement PATCH /api/moderation-events/{moderation_id}/production-locations/{os_id}/ endpoint

### PR DESCRIPTION
**Pull request by mazursasha1990**
[OSDEV-1409](https://opensupplyhub.atlassian.net/browse/OSDEV-1409) - Moderation. Implement PATCH api/moderation-events/{modreation_id}/production-locations/{os_id}/ endpoint.

In this PR was introduced a new `PATCH /api/moderation-events/{moderation_id}/production-locations/{os_id}/` endpoint. This endpoint allows the creation of a new contribution for an existing production location based on the provided moderation event.

While executing this endpoint, the `api_facilitylistitem` record is created with the status **"CONFIRMED_MATCH"**. 
The `api_facilitymatch` record is created with the status **"CONFIRMED"**, a confidence value of **1**, and the following result: `{"match_type": "moderation_event_confirmed_match"}`.

This behavior simulates a scenario in which DedupeHub returns a single potential match that is subsequently confirmed manually by the moderator.

[OSDEV-1409]: https://opensupplyhub.atlassian.net/browse/OSDEV-1409?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: doc/release/RELEASE-NOTES.md
Comment (mazursasha1990) [reply doc/release/RELEASE-NOTES.md]: Removed

File: src/django/api/views/v1/moderation_events.py
Comment (mazursasha1990) [reply src/django/api/views/v1/moderation_events.py:129]: Nevertheless, I decided to change this method in a different way to adhere to the Single Responsibility Principle.

File: src/django/api/tests/test_moderation_events_add_production_location.py
Comment (mazursasha1990) [reply src/django/api/tests/test_moderation_events_add_production_location.py:65]: No, I refactored tests.

File: src/django/api/moderation_event_actions/approval/event_approval_template.py
Comment (mazursasha1990) [reply src/django/api/moderation_event_actions/approval/event_approval_template.py]: Refactored this part of code.

File: src/django/api/views/v1/moderation_events.py
Comment (VadimKovalenkoSNF) [reply src/django/api/views/v1/moderation_events.py:129]: @mazursasha1990 maybe it's a good idea to leave this comment in the code:

`DRF resolves actions based on their url_path and methods. Since both add_production_location and update_production_location have the same url_path, DRF gets confused about which action to invoke which leads to errors.`

**Review by VadimKovalenkoSNF — APPROVED** [review]
Approved. Left minor [suggestion](https://github.com/opensupplyhub/open-supply-hub/pull/438/files#r1883378833), feel free to skip.

File: src/django/api/views/v1/moderation_events.py
Comment (vladsha-dev) [inline src/django/api/views/v1/moderation_events.py:143]: For this part, I recommend implementing something like this

  @detail_route(methods=['get', 'head'], url_path='foo', url_name='foo')
   def foo(self, request, pk=None):
       if request.method == 'get':
           return self.foo_get(request, pk)
       if request.method == 'head':
           return self.foo_head(request, pk)
   def foo_get(self, request, pk):
       pass
   def foo_head(self, request, pk):
       pass
       
I think it will make the code leaner

**Review by vladsha-dev — CHANGES_REQUESTED** [review]

File: src/django/api/views/v1/moderation_events.py
Comment (mazursasha1990) [reply src/django/api/views/v1/moderation_events.py]: I refactored the code using [Routing additional HTTP methods for extra actions](https://www.django-rest-framework.org/api-guide/viewsets/#routing-additional-http-methods-for-extra-actions)
cc @VadimKovalenkoSNF

File: src/django/api/views/v1/moderation_events.py
Comment (vladsha-dev) [inline src/django/api/views/v1/moderation_events.py:39]: I think it can be moved to the method level. I'm not sure we really need it at the class level since it's only used inside the `get_permissions` method.

File: src/django/api/views/v1/moderation_events.py
Comment (vladsha-dev) [inline src/django/api/views/v1/moderation_events.py:112]: Since this decorator is primarily focused on OpenSearch-related errors or errors related to searching via OpenSearch, I think it makes sense either not to set it that or to create a new handler and rename this one to clearly reflect its purpose. One of the errors mentions input, which in the case of PATCH requests may be unusual to receive.

File: src/django/api/views/v1/moderation_events.py
Comment (vladsha-dev) [inline src/django/api/views/v1/moderation_events.py:134]: Since this decorator is primarily focused on OpenSearch-related errors or errors related to searching via OpenSearch, I think it makes sense either not to set it that or to create a new handler and rename this one to clearly reflect its purpose. One of the errors mentions input, which in the case of PATCH requests may be unusual to receive.

File: src/django/api/views/v1/moderation_events.py
Comment (vladsha-dev) [inline src/django/api/views/v1/moderation_events.py:157]: Since this decorator is primarily focused on OpenSearch-related errors or errors related to searching via OpenSearch, I think it makes sense either not to set it that or to create a new handler and rename this one to clearly reflect its purpose. One of the errors mentions input, which in the case of PATCH requests may be unusual to receive.

File: src/django/api/views/v1/moderation_events.py
Comment (vladsha-dev) [inline src/django/api/views/v1/moderation_events.py:170]: This kind of validation can be moved to a regular Django serializer, whose purpose is validation.

File: src/django/api/views/v1/moderation_events.py
Comment (vladsha-dev) [inline src/django/api/views/v1/moderation_events.py:164]: This kind of validation can be moved to a regular Django serializer, whose purpose is validation.

**Review by vladsha-dev — CHANGES_REQUESTED** [review]
1. Since the endpoint is now `/api/moderation-events/{moderation_id}/production-locations/**{os_id}/**`, I think it would be helpful to update the RELEASE-NOTES.md, as well as the description and title of the PR, including the Jira task description and title with the new endpoint name
2. Left comments in the files

File: src/django/api/views/v1/moderation_events.py
Comment (roman-stolar) [reply src/django/api/views/v1/moderation_events.py:45]: +1

File: src/django/api/views/v1/moderation_events.py
Comment (roman-stolar) [reply src/django/api/views/v1/moderation_events.py:110]: +1

**Review by roman-stolar — COMMENTED** [review]
In general LGTM, the Template pattern suits our requirements, good job!
Agree with @vladsha-dev comments.
Will approve when @vladsha-dev will be addressed.

File: src/django/api/tests/base_moderation_events_production_location_test.py
Comment (roman-stolar) [reply src/django/api/tests/base_moderation_events_production_location_test.py:280]: Incorrect suggestion.

File: src/django/api/tests/test_moderation_events_update.py
Comment (roman-stolar) [reply src/django/api/tests/test_moderation_events_update.py:75]: good suggestion

**Review by vladsha-dev — CHANGES_REQUESTED** [review]

File: src/django/api/constants.py
Comment (mazursasha1990) [reply src/django/api/constants.py]: Created a new class `APIV1ModerationEventErrorMessages`.

File: src/django/api/permissions.py
Comment (mazursasha1990) [reply src/django/api/permissions.py]: Changed it to lowercase.

File: src/django/api/views/v1/moderation_events.py
Comment (mazursasha1990) [reply src/django/api/views/v1/moderation_events.py]: Good notice.
Moved it to the get_permissions method.
Thanks.

File: src/django/api/views/v1/moderation_events.py
Comment (mazursasha1990) [reply src/django/api/views/v1/moderation_events.py]: Removed `handle_errors_decorator` from this action.

File: src/django/api/views/v1/moderation_events.py
Comment (mazursasha1990) [reply src/django/api/views/v1/moderation_events.py]: Removed `handle_errors_decorator` from this action.

File: src/django/api/views/v1/moderation_events.py
Comment (mazursasha1990) [reply src/django/api/views/v1/moderation_events.py]: Removed `handle_errors_decorator` from this action.

File: src/django/api/tests/test_moderation_events_update.py
Comment (mazursasha1990) [reply src/django/api/tests/test_moderation_events_update.py]: Changes in this file were added accidentally. I reverted them. But I removed an unnecessary print statement.

File: src/django/api/tests/test_moderation_events_update.py
Comment (mazursasha1990) [reply src/django/api/tests/test_moderation_events_update.py]: Changes in this file were added accidentally. I reverted them.

File: src/django/api/tests/base_moderation_events_production_location_test.py
Comment (mazursasha1990) [reply src/django/api/tests/base_moderation_events_production_location_test.py]: Fixed

File: src/django/api/moderation_event_actions/approval/event_approval_template.py
Comment (mazursasha1990) [reply src/django/api/moderation_event_actions/approval/event_approval_template.py:30]: I added additional context to the docstring. 
Please check it.

File: src/django/api/moderation_event_actions/approval/event_approval_template.py
Comment (mazursasha1990) [reply src/django/api/moderation_event_actions/approval/event_approval_template.py:39]: Used @transaction.atomic decorator.

File: src/django/api/moderation_event_actions/approval/event_approval_template.py
Comment (mazursasha1990) [reply src/django/api/moderation_event_actions/approval/event_approval_template.py:148]: Removed an empty line.

File: src/django/api/moderation_event_actions/approval/event_approval_template.py
Comment (mazursasha1990) [reply src/django/api/moderation_event_actions/approval/event_approval_template.py:250]: Changed.

File: src/django/api/tests/test_moderation_events_add_production_location.py
Comment (mazursasha1990) [reply src/django/api/tests/test_moderation_events_add_production_location.py:30]: It's okay to use here 'post' request.

File: src/django/api/moderation_event_actions/approval/event_approval_template.py
Comment (mazursasha1990) [reply src/django/api/moderation_event_actions/approval/event_approval_template.py]: Changed lox prefix to `[API V1 Location Contribution Approval]`.

**Comment by VadimKovalenkoSNF** [comment]
@mazursasha1990 pls, update title of the PR to reflect new path parameter presence in the endpoint.

**Review by roman-stolar — DISMISSED** [review]
LGTM, as I see all comments created by @vladsha-dev was addressed. @vladsha-dev please resolve the conversation.

**Comment by roman-stolar** [comment]
@mazursasha1990 Please fix unit tests.

**Review by roman-stolar — APPROVED** [review]
Re-Approved

**Review by VadimKovalenkoSNF — APPROVED** [review]
Looks good, but let's wait for @vladsha-dev for review.

**Review by vladsha-dev — APPROVED** [review]
LGTM

**Pull request merged**
